x/stake/keeper/val_state_change.go

```diff
@@ -0,0 +1,287 @@
+package keeper
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+//_________________________________________________________________________
+// Accumulated updates to the active/bonded validator set for tendermint
+
+// get the most recently updated validators
+//
+// CONTRACT: Only validators with non-zero power or zero-power that were bonded
+// at the previous block height or were removed from the validator set entirely
+// are returned to Tendermint.
+func (k Keeper) GetTendermintUpdates(ctx sdk.Context) (updates []abci.ValidatorUpdate) {
+
```

Comment:
Logic here looks sound and is easy to follow along. It is quite long though. Mind I suggest we break it up a bit? At least an auxiliary function to get the old validators by power (i.e. what you store in `last`).

Reply:
Yup, split out functions to load the old validator set and sort the validators to be removed.